server/src/test/java/org/elasticsearch/bootstrap/ElasticsearchTests.java

```diff
@@ -0,0 +1,71 @@
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.bootstrap;
+
+import org.elasticsearch.entitlement.runtime.policy.InboundNetworkEntitlement;
+import org.elasticsearch.entitlement.runtime.policy.LoadNativeLibrariesEntitlement;
+import org.elasticsearch.entitlement.runtime.policy.OutboundNetworkEntitlement;
+import org.elasticsearch.entitlement.runtime.policy.Policy;
+import org.elasticsearch.entitlement.runtime.policy.Scope;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Map.entry;
+import static org.hamcrest.Matchers.aMapWithSize;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+public class ElasticsearchTests extends ESTestCase {
+    public void testFindPluginsWithNativeAccess() {
+
+        var policies = Map.ofEntries(
+            entry(
+                "plugin-with-native",
+                new Policy(
+                    "policy",
+                    List.of(
+                        new Scope("module.a", List.of(new LoadNativeLibrariesEntitlement())),
+                        new Scope("module.b", List.of(new InboundNetworkEntitlement()))
+                    )
+                )
+            ),
+            entry(
+                "another-plugin-with-native",
+                new Policy(
+                    "policy",
+                    List.of(
+                        new Scope("module.a2", List.of(new LoadNativeLibrariesEntitlement())),
+                        new Scope("module.b2", List.of(new LoadNativeLibrariesEntitlement())),
+                        new Scope("module.c2", List.of(new InboundNetworkEntitlement()))
+
+                    )
+                )
+            ),
+            entry(
+                "plugin-without-native",
+                new Policy(
+                    "policy",
+                    List.of(
+                        new Scope("module.a3", List.of(new InboundNetworkEntitlement())),
+                        new Scope("module.b3", List.of(new OutboundNetworkEntitlement()))
+                    )
+                )
+            )
+        );
+
+        var pluginsWithNativeAccess = Elasticsearch.findPluginsWithNativeAccess(policies);
+
+        assertThat(pluginsWithNativeAccess, aMapWithSize(2));
+        assertThat(pluginsWithNativeAccess.keySet(), containsInAnyOrder("plugin-with-native", "another-plugin-with-native"));
+        assertThat(pluginsWithNativeAccess.get("plugin-with-native"), containsInAnyOrder("module.a"));
+        assertThat(pluginsWithNativeAccess.get("another-plugin-with-native"), containsInAnyOrder("module.a2", "module.b2"));
```

Comment:
This set of assertions won't fail if, for example, `pluginsWithNativeAccess.get("plugin-with-native")` contains extra bonus modules. (These kinds of assertion sequences that check one property at a time can be error prone, and we have to remember to maintain them as the data structure evolves, or else we end up with incomplete testing.)

We should construct the expected map and then use a single `assertEquals(expected, actual)` if possible.

Reply:
Actually they do; `contains` and `containsInAnyOrder` are a bit surprising that way: they match all of the elements. 
From the javadoc: 
> For a positive match, the examined iterable must be of the same length as the number of specified matchers

Reply:
Ah my mistake. So this sequence of code is actually equivalent to `assertEquals`. In that case, I think `assertEquals` is clearer; these advanced assertions are best left for tricky cases where `assertEquals` doesn't apply.

Reply:
(which makes the `aMapWithSize(2)` redundant - I'll remove it)